org.lflang/src/org/lflang/generator/DockerGeneratorBase.java

```diff
@@ -124,92 +147,103 @@ public void addFederate(
-                dockerFilePath.toFile().delete();
-            }
-            var contents = (String) k.get(Key.DOCKER_FILE_CONTENT);
-            FileUtil.writeToFile(contents, dockerFilePath);
-            System.out.println(getDockerBuildCommand(
-                lfModuleName, dockerFilePath,
-                dockerComposeFilePath.getParent(),
-                (String) k.get(Key.DOCKER_COMPOSE_SERVICE_NAME)));
+        for (DockerData dockerData : dockerDataList) {
+            writeDockerFile(dockerData);
+            System.out.println(
+                getDockerBuildCommandMsg(
+                    dockerComposeFilePath, dockerData));
         }
 
+        System.out.println(getDockerComposeUpMsg(dockerComposeFilePath));
         if (isFederated && host != null) {
             appendRtiToDockerComposeServices(
-                composeServices,
                 "lflang/rti:rti",
-                host,
-                nFederates
+                host
             );
         }
-        writeFederatesDockerComposeFile(dockerComposeFilePath, composeServices, "lf");
+        writeFederatesDockerComposeFile(dockerComposeFilePath, "lf");
     }
 
     /**
-     * A template function for generating target-specific docker file content.
+     * Writes the docker file given the docker data.
      *
-     * @param lfModuleName The name of the LF module currently generating.
+     * @param dockerData The docker data as specified in the DockerData class.
      */
-    protected String generateDockerFileContent(
-        String lfModuleName
-    ) {
-        throw new UnsupportedOperationException("Docker file content is not implemented");
+    private void writeDockerFile(DockerData dockerData) throws IOException {
+        var dockerFilePath = dockerData.getFilePath();
+        if (dockerFilePath.toFile().exists()) {
+            dockerFilePath.toFile().delete();
+        }
+        FileUtil.writeToFile(dockerData.getFileContent(), dockerFilePath);
     }
 
     /**
      * Write a Dockerfile for the current federate as given by filename.
-     * @param The directory where the docker compose file is generated.
-     * @param The name of the docker file.
-     * @param The name of the federate.
      */
-    public String getDockerComposeCommand() {
+    private String getDockerComposeCommand() {
         String OS = System.getProperty("os.name").toLowerCase();
         return (OS.indexOf("nux") >= 0) ? "docker-compose" : "docker compose";
     }
 
+    /**
+     * Throws a Runtime Exception if any field in "dockerData" is null.
+     */
+    private void validateNotNull(DockerData dockerData) {
+        if (dockerData.getFilePath() == null || dockerData.getFileContent() == null ||
+            dockerData.getComposeServiceName() == null ||
+            dockerData.getBuildContext() == null ||
+            dockerData.getContainerName() == null) {
+                throw new RuntimeException("Fields in DockerData cannot be null");
```

Comment:
Based on the logic in this function, it appears that `filePath` or `fileContent` may be `null`. Either the message has to be made more specific (e.g., `filePath` or `fileContent` field must be set) or more generic (e.g., Missing fields in `DockerData` instance).

Reply:
good point. I will change the error message to what you suggested.